distanceDiracGaussianOPTw.m

```diff
@@ -0,0 +1,296 @@
+function [D,G] = distanceDiracGaussianOPTw(xw,N,L)
+%% 
+% distanceDiracGaussian.m calculates the distance between dirac mixture
+% approximation and normal distribution in terms of Localized Cumulative
+% Distribution. Distance consists of three terms: D1,D2 and D3.
+% Detailed inference can be found in:
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+% "Efficient Deterministic Dirac Mixture Approximation of Gaussian
+% Distributions", Igor Gilitschenski and Uwe D. Hanebeck
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+%
+% Parameters:
+%   x: location of dirac distributions
+%   N: dimension of dirac mixture approximation
+%   L: number of component for each dimension 
+%
+% Return values:
+%   D: distance between dirac mixture approximation and normal distribution
+%   G: gradient of D
+%
+% History:
+% * 2018/01/04 Dantong Wang
+
+%% weight and range for integral
+x = reshape(xw(1:N*L),[N,L]);
+w = reshape(xw(N*L+1:L+N*L),[1,L]);
+bmax = 3;
+%% compute the exponential interation used in subroutines
+c = sum(x.^2,1);
+divcb = -c./(2+4*bmax^2);
+divc0 = -c./2;
+ex_inb = ei(divcb);
+ex_in0 = ei(divc0);
+
+%% calculate D1
+D1 = pi^(N/2)*dmd_IN(N,bmax);
+
+%% calculate D2
+k = floor(N/2);
+l = N - k*2;
+% if N is even
+if l == 0
+    D2 = (2*pi)^(N/2)*sum(w.*(dmd_Jmm(k,bmax,c,divcb,ex_inb)-dmd_Jmm(k,0,c,divc0,ex_in0)),2);
+% if N is odd
+else
+    D2 = integral(@(b)dmd_D2Integral(N,x,b,w),0,3);
+end
+% calclate Tij as a matrix
+Tij = permute(sum(bsxfun(@minus,x,permute(x,[1,3,2])).^2,1),[2,3,1])+1e-5;
+
+%% calculate D3
+%calculate exponential interation of Tij
+en_inT = ei(-1/2*(Tij./(2*bmax^2)));
+D31 = bmax^2/2*exp(-1/2*(Tij./(2*bmax^2)));
+D32 = (Tij/8).*en_inT;
+D3 = pi^(N/2)*sum(sum(bsxfun(@times,w',bsxfun(@times,w,(D31+D32))),1),2);
+
+%% Gradient
+%% calculate G1
+% if N is even
+if l == 0
+    G1 = 2*(2*pi)^(N/2)*(w.*x.*(dmd_Jkk1(k,bmax,c,divcb,ex_inb)-dmd_Jkk1(k,0,c,divc0,ex_in0)));
+% if N is odd
+else
+    G1 = 2*(2*pi)^(N/2)*w.*x.*integral(@(b)dmd_G1Integral(N,x,b),0,3,'ArrayValue',true);
+end
+
+%% calculate G2
+%calculate the sigma in G2
+SIGMA = permute(w'.*sum(w.*bsxfun(@minus,permute(x,[2,3,1]),permute(x,[3,2,1])).*en_inT,2),[3,1,2]);
+G2 = pi^(N/2)/2.*SIGMA;
+
+%% calculate Gw
+%calculate dD2dw
+if l == 0
+    dD2dw = (2*pi)^(N/2)*(dmd_Jmm(k,bmax,c,divcb,ex_inb)-dmd_Jmm(k,0,c,divc0,ex_in0));
+else
+    dD2dw = integral(@(b)dmd_dD2dwIntegral(N,x,b,w),0,3,'ArrayValue',true);
+end
+%calculate dD3dw
+dD3dw = pi^(N/2)*(sum(bsxfun(@times,w',(D31+D32)),1) + sum(bsxfun(@times,w,(D31+D32)),2)');
+Gw = -2*dD2dw + dD3dw;
+%% output
+D = D1-2*D2+D3;
+G = [reshape(G1+G2,[1,N*L]),Gw];
+end
+%%
+function D2Int = dmd_D2Integral(N,x,b,w)
```

Comment:
in this file, there are a few functions starting with "dmd". should these be "cmd"?